src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,216 @@
+            Range requestedRange,
+            [NotNullWhen(true)] out (Range Range, List<int> Tokens)? cachedTokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache) ||
+                !documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                // No cached results found
+                cachedTokens = null;
+                return false;
+            }
+
+            var tokens = new List<int>();
+
+            // Keep track of where the cached range begins
+            Position? cachedRangeStart = null;
+
+            // Keep track of where the cached range ends
+            var cachedRangeNumLines = 0;
+
+            // Keeps track of the current empty line count (needed for calculating line offsets)
+            var emptyLineCount = 0;
+
+            // Retrieve the cached tokens associated with the passed-in range
+            for (var line = requestedRange.Start.Line; line < requestedRange.End.Line; line++)
+            {
+                if (!lineToTokensDict.TryGetValue(line, out var lineTokens))
+                {
+                    if (tokens.Count != 0)
+                    {
+                        // We already have some sort of cached results, but we've now reached
+                        // a gap in the cache which we have no tokens for.
+                        // Since we only want to return a continuous range, we'll exit early.
+                        break;
+                    }
+
+                    continue;
+                }
+
+                if (cachedRangeStart is null)
+                {
+                    cachedRangeStart = new Position { Line = line, Character = 0 };
+                }
+
+                cachedRangeNumLines++;
+
+                // If a line's List is empty, it means the line has no tokens
+                if (lineTokens.Count == 0)
+                {
+                    emptyLineCount++;
+                    continue;
+                }
+
+                // We've found a non-empty cached line. Go through and process it.
+                for (var lineTokenIndex = 0; lineTokenIndex < lineTokens.Count; lineTokenIndex++)
+                {
+                    // If we're at the start of the cached range, the line offset should be relative
+                    // to the start of the document.
+                    if (tokens.Count == 0)
+                    {
+                        tokens.Add(line);
+                    }
+                    // If we're not at start at the cached range but we're at the start of a line,
+                    // the line offset should be relative to the end of the last cached line.
+                    else if (lineTokenIndex == 0)
+                    {
+                        tokens.Add(1 + emptyLineCount);
+                    }
+                    else
+                    {
+                        tokens.Add(lineTokens[lineTokenIndex]);
+                    }
+                }
```

Comment:
ok.  it took me a lot of time to understand what was going on here.  But i think i somewhat get it (though i still don't get certain bits of internal data in the loop).  You are basically taking all the tokens on the line, and using them *as is*, *except* for the *Very first int data* on the line, which indicates what hte starting line-offset is.

I would rewrite this then.  instead of one loop that handles everything.  I would simply have special logic for the first item.  and then do an AddRange of all the data past the first item.